### ISSUE TITLE
integrate zod into Slonik

### ISSUE THREAD

**Issue by gajus** [gajus/slonik#364]
At the moment, we can provide type suggestions the following way:

```ts
const persons = await pool.any<{id: number, name: string}>(sql`
  SELECT id, name
  FROM person
`);
```

While this works, there are no guarantees that this is what the query actually returns.

In order to add more excellent type safety, including runtime safety, Slonik is going to adopt [zod](https://www.npmjs.com/package/zod) natively.

The current thinking is that the API is going to be:

```ts
const persons = await pool.any(sql`
  ${z.object({
    id: z.number(),
    name: z.string(),
  })}
  SELECT id, name
  FROM person
`);
```

The reason for adding `z.object` to the `sql` template itself is because, ideally, we want them to live together.

We can then toggle runtime validation to ensure that queries return exactly what we type them to return, i.e. If this query were to change to:

```ts
const persons = await pool.any(sql`
  ${z.object({
    id: z.number(),
    name: z.string(),
  })}
  SELECT id, name, description
  FROM person
`);
```

We would throw an error about the unknown `description` property.

In contrast to build-time type safety, this mechanism of validation ensures that the application continues to work as expected (or breaks loudly) if database schema drifts from whatever we typed at build time, e.g. In the earlier example, we may have assumed that `name` is non-nullable based on our current knowledge of schema, but if that changed during the lifetime of the application, we'd want to know about it.

**Comment by gajus** [comment]
High-level, it would look something like this:

```ts
const myTag = <T extends ZodTypeAny>(strings: TemplateStringsArray, maybeZodObject: T): z.infer<T> => {
  console.log('maybeZodObject', maybeZodObject);

  return strings.join(',') as any;
};

const foo = myTag`
  ${z.object({
    name: z.string(),
  })}
Hello, World!`;

console.log(foo);
```

**Comment by mmkal** [comment]
Interesting! Re this statement:

>While this works, there are no guarantees that this is what the query actually returns.

I know you've used `@slonik/typegen` in the past for this - would you be interested in working together on something like this? There are overlapping goals here I think. In fact `@slonik/typegen` is designed to ensure that you _can_ guarantee it's what the query actually returns - by relying on Postgres, which already has a fully type system. The aim is to achieve that _without_ having to duplicate the types manually, which could be laborious and error-prone. 

I'd love to discuss options here, I have a few ideas on how it could go further in achieving type safety that might require updates like what you describe here.

**Comment by gajus** [comment]
I was already thinking of how to combine this with `@slonik/typegen`.

I think the move is to allow `@slonik/typegen` to generate zod objects for queries.

The existing API of type suggestions, i.e.

```ts
sql<{id: number}>``;
```

is going to be removed (it clashes with zod logic) in favor of fully embracing zod.

My thinking is that (given that typegen cannot always generate perfect types), we use comment conventions to guide when we expect assistance, e.g.

```ts
const personQuery = sql`
  SELECT id, name
  FROM person
`
```

In cases where typegen is confident, this should automatically become:

```ts
const personQuery = sql`
  ${z.object({
    id: z.number(),
    name: z.string(),
  })}
  SELECT id, name
  FROM person
`
```

**Comment by mmkal** [comment]
> The existing API of type suggestions, i.e.
> 
> ```ts
> sql<{id: number}>``;
> ```
>
> is going to be removed (it clashes with zod logic) in favor of fully embracing zod.

now that the api will be `` sql.type(MyZodObj)`select 123` `` no need to take the existing api away right?

As the maintainer of slonik-typegen I'd love to chat about this more. I'd be very happy to incorporate zod, but my first instinct is that it should be an option, not The Only Way. I think there will always be use cases where we don't want to define a validator that reflects an existing query, and perform the checks at runtime - but we _do_ want types for the query. There has already been some great work done by @janpaepke to augment imperfect outputs of slonik-typegen: https://github.com/mmkal/slonik-tools/tree/master/packages/typegen#enhancing-return-types

If you're open to chatting I'm on discord mmkal#1574 or various slacks (or can do back and forth via GitHub). Let me know!

**Comment by janpaepke** [comment]
In general I think it's a good idea to make sql statements "runtime-safe".
My plans for adding features to typegen was mostly to make it "smarter" in that it could for example also understand functions like coalesce or json_agg.

I tend to think that there should not be two different interfaces to define types, since sync issues are likely. There should be one "source of truth".

From a glance it does feel though, that making the types part of the literal might make it trickier to parse but I'd have to learn more about zod to form an opinion.

I'm open to a chat or video call, if you like.

**Comment by gajus** [comment]
> I think there will always be use cases where we don't want to define a validator that reflects an existing query, and perform the checks at runtime - but we do want types for the query.

I have not thought of yet how, but there will be a method to disable/enable selectively when to run the checks, e.g. only in development mode or sampling.

However, I do think that allowing more than one way to define types is going to result in a mess. Therefore, my current thinking is that even if user does not plan to use runtime validation, types should be defined using the same zod interface.

**Comment by mmkal** [comment]
With the new `` sql.type(...)` ... ` `` API there is only one way - the existing generic way. `sql.type(...)` is just a normal function which provides that generic param. It doesn't even need to be coupled to zod:

```ts
const sql = <T>(...args): Query<T> => ...

type Parser<T> = {parse(input: unknown): T}

const getSqlFromTypeParser => <T>(parser: Parser<T>) => {
  return (...args) => ({
    ...sql<T>(...args),
    parser,
  })
}

sql.type = getSqlFromTypeParser
```

The signature of `sql` doesn't need to change at all, and that we it hasnt forced every single user of slonik to define zod definitions for every query. You use with zod pretty naturally:

```ts
import {z} from 'zod'

const Foo = z.object({ foo: z.string() })

await pool.query(sql.type(Foo)`
  select 'abc' as foo
`
```

or without zod:

```ts
type Foo = { foo: z.string }

await pool.query(sql<Foo>`
  select 'abc' as foo
`
```

or with some other runtime validation library (all it needs is a `parse` method which asserts the input value has the desired shape):

```ts
import {t} from 'some-other-library'

const Foo = z.type({ foo: z.string })

await pool.query(sql.type(Foo)`
  select 'abc' as foo
`
```

**Comment by janpaepke** [comment]
@mmkal could you elaborate why you think it's beneficial to have two ways of defining types, one compiletime and one runtime (and implicitly compile time)?

I tend to agree with @gajus that two different syntaxes with subtle differences is a recipe for messiness.

What I do think I better though is to get the type definition separated from the query itself, as you suggest with sql.type.

**Comment by gajus** [comment]
@mmkal I mostly got it to work in a backwards compatible way https://github.com/gajus/slonik/pull/369

**Comment by gajus** [comment]
> @mmkal could you elaborate why you think it's beneficial to have two ways of defining types, one compiletime and one runtime (and implicitly compile time)?

Technically, it isn't adding a second way of adding types. `sql.type` just constructs a bound `sql<Type>`.

**Comment by mmkal** [comment]
> > @mmkal could you elaborate why you think it's beneficial to have two ways of defining types, one compiletime and one runtime (and implicitly compile time)?
> 
> Technically, it isn't adding a second way of adding types. `sql.type` just constructs a bound `sql<Type>`.

Agreed. If you look at #371 it's now a non harmful change. 

However, I do think the two sources of truth point is a very valid concern. Postgres is the real source of truth, so I think if slonik-typegen gets really good at parsing queries the need for zod validations will be rare. Which is why I feel strongly that this change should be a pure, non-breaking feature add (which is what I aimed for in #371).

**Comment by gajus** [comment]
Released in [v30.1.0](https://github.com/gajus/slonik/releases/tag/v30.1.0)

https://contra.com/p/gkOQlbLq-validating-postgre-sql-query-results-using-runtime-checks

**Comment by hjr3** [comment]
Huge fan of this change. I was doing this myself with a small wrapper around slonik. Glad to see this is part of the interface.
